include/dmlc/intrusive_ptr.h

```diff
@@ -0,0 +1,259 @@
+/*!
+ * Copyright (c) by Contributors 2020
+ * \file intrusive_ptr.h
+ * \brief Implementation of Intrusive Ptr.
+ */
+#ifndef DMLC_INTRUSIVE_PTR_H_
+#define DMLC_INTRUSIVE_PTR_H_
+
+#include <atomic>
+#include <cinttypes>
+#include <functional>
+
+namespace dmlc {
+/*!
+ * \brief Helper class for embedding reference counting into client objects.
+ */
+class IntrusivePtrCell {
+  std::atomic<int32_t> count_;
+  template <typename T> friend class IntrusivePtr;
+
+  std::int32_t Inc() noexcept { return ++count_; }
```

Comment:
Inc -> IncRef, consider add memory order requirement as in https://github.com/apache/incubator-tvm/blob/master/include/tvm/runtime/object.h#L772